Add various bounds-related simplifier rules

rootjalex opened the pull request:
All of these simplifier rules came from `find_constant_bounds` failures. Some are oddly specific and I can be convinced to remove them (i.e. in Simplify_Max.cpp and Simplify_Select.cpp), but many of them are useful for `find_constant_bounds`.

rootjalex commented:
I should have noted: all rules have been formally verified and confirmed to follow the reduction order

abadams commented on src/Simplify_Add.cpp:
```diff
@@ -74,6 +74,7 @@ Expr Simplify::visit(const Add *op, ExprInfo *bounds) {
              rewrite(select(x, y, z) + (w + select(x, u, v)), select(x, y + u, z + v) + w) ||
              rewrite(select(x, y, z) + (select(x, u, v) - w), select(x, y + u, z + v) - w) ||
              rewrite(select(x, y, z) + (w - select(x, u, v)), select(x, y - u, z - v) + w) ||
+             rewrite(select(x, c0 - y, c1) + c2, select(x, fold(c0 + c2) - y, fold(c1 + c2)), !overflows(c0 + c2) && !overflows(c1 + c2)) ||
```
Rules fail to match if a fold on the RHS overflows, so the predicate is unnecessary

abadams commented on src/Simplify_Add.cpp:
```diff
@@ -148,6 +150,7 @@ Expr Simplify::visit(const Add *op, ExprInfo *bounds) {
                rewrite(max(y + c0, x) + c1, max(y, x + c1), c0 + c1 == 0) ||
                rewrite(max(x, y) + min(x, y), x + y) ||
                rewrite(max(x, y) + min(y, x), x + y) ||
+               rewrite(max(x, y*c0 + z) + (u - y)*c0, max(x - y*c0, z) + u*c0) ||
```
Why does this rule show up twice?

rootjalex replied:
copy-paste error, will fix

abadams commented on src/Simplify_Max.cpp:
```diff
@@ -97,6 +97,10 @@ Expr Simplify::visit(const Max *op, ExprInfo *bounds) {
              rewrite(max(max(x, y), min(z, x)), a) ||
              rewrite(max(max(x, y), min(z, y)), a) ||
 
+             rewrite(max(select(x, max(z, y), w), z), max(select(x, y, w), z)) ||
```
Worth adding the versions where the else case of the select is the redundant max?

rootjalex replied:
I don't think I saw these in logging, but it can't hurt. Will add

abadams commented on src/Simplify_Max.cpp:
```diff
@@ -97,6 +97,10 @@ Expr Simplify::visit(const Max *op, ExprInfo *bounds) {
 
+             rewrite(max(select(x, max(z, y), w), z), max(select(x, y, w), z)) ||
+             rewrite(max(select(x, max(z, y), w), y), max(select(x, z, w), y)) ||
+             rewrite(max(min(x*c0, y + c1) - min(select(0 < x, y, min(x*c0, y + c1)), x*c0), 0), 0, c1 < 0) ||
```
This one is probably too specific

abadams commented on src/Simplify_Min.cpp:
```diff
@@ -97,6 +97,10 @@ Expr Simplify::visit(const Min *op, ExprInfo *bounds) {
              rewrite(min(max(x, y), min(z, y)), b) ||
 
+             rewrite(min(max(x, y + c0), y), y, c0 > 0) ||
+             rewrite(min(select(x, min(z, y), w), y), min(select(x, z, w), y)) ||
```
Worth adding versions where the min is in the else case?

abadams commented on src/Simplify_Mul.cpp:
```diff
@@ -79,6 +79,12 @@ Expr Simplify::visit(const Mul *op, ExprInfo *bounds) {
             rewrite(max(x, y) * min(x, y), x * y) ||
             rewrite(max(x, y) * min(y, x), y * x) ||
+
+            rewrite(min(x * c0, c1) * c2, min(x * fold(c0 * c2), fold(c1 * c2)), c2 > 0) ||
```
I'm not comfortable with rules that makes constants larger, because it risks triggering signed integer overflow, and we're not completely sound with respect to that.

rootjalex replied:
In one of the above comments, you mentioned that a rule will fail to match if it detects overflow - so is this specifically not wanting to produce larger constants with multiplication? Regardless, I removed these in a827ae5

rootjalex replied:
It also appears that the second of these rules was causing the fuzz_simplify failure on the buildbots of 0a00af4 , so the lack of comfort with these rules seems plenty reasonable.

rootjalex commented:
@abadams I believe I have addressed all of your comments - I did not see the min-select and max-select variants in my logging but they do seem useful, so I added them as you requested.
